kythe/typescript/indexer.ts

```diff
@@ -30,6 +30,25 @@ const LANGUAGE = 'typescript';
  * code duplication.
  */
 export interface IndexerHost {
+  /**
+   * Computes the VName (and signature) of a ts.Symbol. A Context can be
+   * optionally specified to help disambiguate nodes with multiple declarations.
+   * See the documentation of Context for more information.
+   */
+  getSymbolName(sym: ts.Symbol, ns: TSNamespace, context?: Context): VName;
+  /**
+   * Sets the VNames of a ts.Symbol. Throws if a VName was already set.
+   */
+  setSymbolNames(sym: ts.Symbol, names: [VName|null, VName|null, VName|null]):
```

Comment:
This could still lead to symbol pollution, as we discussed yesterday. I think the indexing host should be source of truth, with only unidirectional communication from the host to the indexer. `getSymbolName` can cache symbols.

Reply:
There needs to be a way to alias symbols, because that is what the `Visitor` currently does.
How do you feel about the `aliasSymbol` method that I replaced this with? I can also add a `throw` if you try to alias an already set `Symbol`, although I'm not sure that's necessary.

Reply:
How do you feel about keeping a set of aliased symbols in the Visitor itself that we hit before going to `getSymbolNames`? This would keep the aliasing local to the TypeScript indexer itself.

If this is too much work (or too nitpicky) I think it's okay, as we can move that later and we will probably refactor `visitImport` where this method is used; if I remember correctly Evan wasn't always pleased with the way that imports are handled.

Do you expect to be using `aliasSymbol` for the TS->proto work?

Reply:
I think we want the aliased symbols to be in the symbolNames table, so that they are available (by their aliases) to downstream plugins.
I don't expect to use `aliasSymbol` for the TS->proto work, but I do expect to consume symbols that are aliased, so I don't want the aliases to be scoped to a `Visitor`.
I considered subclassing `IndexerHost` to add `aliasSymbol` and make it only visible to `Visitor`, but I thought that was too complex.

Reply:
Let's leave this conversation open in case Evan has some comments, but this seems okay to me for now.

Reply:
How do you feel about moving the `visitImport` logic into `getSymbolName`? We can do it in a separate PR. I'm taking a look at it implementing it now to see if it's a reasonable option.

Reply:
i.e. so the aliasing is all done within the host, and the symbolNames can be completely contained?

Reply:
yep.

Reply:
I feel good about that, but I think it belongs in a separate PR. This one really just moves things around without changing much logic.

Reply:
LGTM